collector/lib/HostHeuristics.cpp

```diff
@@ -185,13 +186,74 @@ class GardenLinuxHeuristic : public Heuristic {
   }
 };
 
+class EBPFPermissionsHeuristic : public Heuristic {
+  // On some platforms, SELinux is known to cause some problems with map creation.
+  // This heuristic will check whether map creation is possible, and if not it
+  // will switch to Kernel Module based collection, logging this as a warning.
+  //
+  // While this issue is known to affect RHEL-7, the heuristic is platform
+  // agnostic to ensure coverage of all platforms.
+  void Process(HostInfo& host, const CollectorConfig& config, HostConfig* hconfig) const {
+    if (!config.UseEbpf() || !host.HasEBPFSupport()) {
+      return;
+    }
+
+    CLOG(DEBUG) << "Attempting to create eBPF map...";
+
+    if (canCreateMap()) {
+      CLOG(DEBUG) << "eBPF map creation successful.";
+      return;
+    }
+
+    CLOG(WARNING) << "Unable to create eBPF map, switching to KernelModule collection";
+    CLOG(WARNING) << "To enable eBPF, check your SELinux settings.";
+    hconfig->SetCollectionMethod("kernel-module");
+  }
+
+ private:
+  bool canCreateMap() const {
+    int map_fd = 0;
+    union bpf_attr attr = {0};
+
+    // it doesn't matter what type we use here, except for 0 (which is invalid)
+    attr.map_type = BPF_MAP_TYPE_HASH;
+    attr.key_size = sizeof(uint64_t);
+    attr.value_size = sizeof(uint64_t);
+    attr.max_entries = 1;
+    attr.map_flags = 0;
+
+    if ((map_fd = sys_bpf(BPF_MAP_CREATE, &attr, sizeof(attr))) < 0) {
```

Comment:
I supposed that closing the fd below will release all the resources ?

Reply:
yeah, when you close the file descriptor the map is deallocated (unless it is pinned, which this one of course isn't :) )